src/link/Elf/relocatable.zig

```diff
@@ -418,16 +419,21 @@ fn writeSyntheticSections(elf_file: *Elf) !void {
     }
     if (elf_file.section_indexes.eh_frame_rela) |shndx| {
         const shdr = slice.items(.shdr)[shndx];
-        const sh_size = math.cast(usize, shdr.sh_size) orelse return error.Overflow;
-        var buffer = try std.ArrayList(u8).initCapacity(gpa, sh_size);
-        defer buffer.deinit();
-        try eh_frame.writeEhFrameRelocs(elf_file, buffer.writer());
-        assert(buffer.items.len == sh_size);
+        const num_relocs = math.cast(usize, @divExact(shdr.sh_size, shdr.sh_entsize)) orelse
+            return error.Overflow;
+        var relocs = try std.ArrayList(elf.Elf64_Rela).initCapacity(gpa, num_relocs);
```

Comment:
It looks like we calculate `num_relocs` to be exactly correct ahead of time. In this case, this buffer should be arena-allocated into the flush arena, and appending should use the non-fallible variants.

Alternately, how I would do it is to delete `calcEhFrameRelocs` to avoid duplicate logic, and use the array list length to calculate the section header data at the **end of flush**. That was the original purpose of flush(), by the way - *only* to write the section headers.

I won't make this a blocking review comment since the previous code has this problem as well.